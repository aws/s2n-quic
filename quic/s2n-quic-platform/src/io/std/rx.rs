// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    buffer,
    message::{
        queue,
        simple::{Message, Ring},
        Message as MessageTrait,
    },
    socket,
};
use s2n_quic_core::io::rx;
use std::io;

impl_io!(Rx);
impl_socket_raw_delegate!(
    impl[Buffer: buffer::Buffer, Socket: socket::raw::AsRaw] Rx<Buffer, Socket>,
    |self| &self.socket
);
impl_socket_mio_delegate!(
    impl[Buffer: buffer::Buffer, Socket: mio::Evented] Rx<Buffer, Socket>,
    |self| &self.socket
);

impl<
        'a,
        Buffer: buffer::Buffer,
        Socket: socket::Simple<Error = io::Error> + socket::Socket<Error = io::Error>,
    > rx::Rx<'a> for Rx<Buffer, Socket>
{
    type Queue = queue::OccupiedWipe<'a, Message>;
    type Error = io::Error;

    fn queue(&'a mut self) -> Self::Queue {
        self.queue.occupied_wipe_mut()
    }

    fn len(&self) -> usize {
        self.queue.free_len()
    }

    fn poll_receive(
        &mut self,
        cx: &mut core::task::Context<'_>,
    ) -> core::task::Poll<io::Result<usize>> {
        let Self { socket, queue } = self;
        socket.poll_receive(cx, |socket| {
            let mut count = 0;
            let mut free = queue.free_mut();

            while let Some(entry) = free.get_mut(count) {
                match socket.recv_from(entry.payload_mut()) {
                    Ok((payload_len, Some(remote_address))) => {
                        entry.set_remote_address(&remote_address);
                        unsafe {
                            // Safety: The payload_len should not be bigger than the number of
                            // allocated bytes.

                            debug_assert!(payload_len < entry.payload_len());
                            let payload_len = payload_len.min(entry.payload_len());

                            entry.set_payload_len(payload_len);
                        }
                        count += 1;
                    }
                    Ok((_payload_len, None)) => {}
                    Err(err) => {
                        if count > 0 && err.kind() == io::ErrorKind::WouldBlock {
                            break;
                        } else {
                            free.finish(count);
                            return Err(err);
                        }
                    }
                }
            }

            free.finish(count);

            Ok(count)
        })
    }
}

#[cfg(test)]
mod tests {
    impl_io_rx_tests!();
}
