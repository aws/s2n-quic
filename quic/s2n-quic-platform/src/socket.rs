// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use cfg_if::cfg_if;

pub mod io;
#[cfg(s2n_quic_platform_socket_mmsg)]
pub mod mmsg;
#[cfg(s2n_quic_platform_socket_msg)]
pub mod msg;
pub mod ring;
pub mod std;

cfg_if! {
    if #[cfg(s2n_quic_platform_socket_mmsg)] {
        pub use mmsg as default;
    } else if #[cfg(s2n_quic_platform_socket_msg)] {
        pub use msg as default;
    } else {
        pub use self::std as default;
    }
}
