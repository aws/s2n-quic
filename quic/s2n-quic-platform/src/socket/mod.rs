#[macro_use]
mod macros;

#[cfg(feature = "std")]
pub mod std;

#[cfg(all(feature = "std", feature = "mio"))]
pub mod mio;

#[cfg(all(feature = "std", feature = "tokio"))]
pub mod tokio;

use s2n_quic_core::inet::SocketAddress;

/// Socket interface for sending simple messages with an address and payload
pub trait Simple {
    type Error;

    /// Receives a payload and returns the length and source address
    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, Option<SocketAddress>), Self::Error>;

    /// Sends a payload to the given address and returns the length of the sent payload
    fn send_to(&self, buf: &[u8], addr: &SocketAddress) -> Result<usize, Self::Error>;
}

/// Provides raw access to os-specific socket handles
pub mod raw {
    use cfg_if::cfg_if;

    cfg_if! {
        if #[cfg(all(target_family = "unix", feature = "std"))] {
            pub type Socket = ::std::os::unix::io::RawFd;

            pub trait AsRaw {
                fn as_raw(&self) -> Socket;
            }

            impl<T: ::std::os::unix::io::AsRawFd> AsRaw for T {
                fn as_raw(&self) -> Socket {
                    self.as_raw_fd()
                }
            }
        } else {
            pub type Socket = *const u8;

            pub trait AsRaw {
                fn as_raw(&self) -> Socket;
            }

            /// make the socket contraints easier to deal with by stubbing out an implementation
            impl<T> AsRaw for T {
                fn as_raw(&self) -> Socket {
                    panic!("raw sockets are not supported on this platform");
                }
            }
        }
    }
}

/// Default sockets for the current platform and enabled features
pub mod default {
    use cfg_if::cfg_if;

    cfg_if! {
        if #[cfg(all(feature = "std", feature = "tokio"))] {
            pub use super::tokio::*;
        } else if #[cfg(all(feature = "std", feature = "mio"))] {
            pub use super::mio::*;
        } else if #[cfg(feature = "std")] {
            pub use super::std::*;
        }
    }
}
