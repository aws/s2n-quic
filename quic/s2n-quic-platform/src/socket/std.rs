// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    buffer::Buffer,
    message::{
        queue,
        simple::{self, Message, Ring},
        Message as _,
    },
};
use errno::errno;
use s2n_quic_core::{event, inet::SocketAddress, path::LocalAddress};

pub use simple::Handle;

pub trait Socket {
    type Error: Error;

    /// Receives a payload and returns the length and source address
    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, Option<SocketAddress>), Self::Error>;

    /// Sends a payload to the given address and returns the length of the sent payload
    fn send_to(&self, buf: &[u8], addr: &SocketAddress) -> Result<usize, Self::Error>;
}

#[cfg(feature = "std")]
impl Socket for std::net::UdpSocket {
    type Error = std::io::Error;

    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, Option<SocketAddress>), Self::Error> {
        debug_assert!(!buf.is_empty());
        let (len, addr) = self.recv_from(buf)?;
        Ok((len, Some(addr.into())))
    }

    fn send_to(&self, buf: &[u8], addr: &SocketAddress) -> Result<usize, Self::Error> {
        debug_assert!(!buf.is_empty());
        let addr: std::net::SocketAddr = (*addr).into();
        self.send_to(buf, &addr)
    }
}

pub trait Error {
    fn would_block(&self) -> bool;
    fn was_interrupted(&self) -> bool;
    fn permission_denied(&self) -> bool;
}

#[cfg(feature = "std")]
impl Error for std::io::Error {
    fn would_block(&self) -> bool {
        self.kind() == std::io::ErrorKind::WouldBlock
    }

    fn was_interrupted(&self) -> bool {
        self.kind() == std::io::ErrorKind::Interrupted
    }

    fn permission_denied(&self) -> bool {
        self.kind() == std::io::ErrorKind::PermissionDenied
    }
}

#[derive(Debug, Default)]
pub struct Queue<B: Buffer>(queue::Queue<Ring<B>>);

impl<B: Buffer> Queue<B> {
    pub fn new(buffer: B) -> Self {
        let queue = queue::Queue::new(Ring::new(buffer, 1));

        Self(queue)
    }

    pub fn free_len(&self) -> usize {
        self.0.free_len()
    }

    pub fn occupied_len(&self) -> usize {
        self.0.occupied_len()
    }

    pub fn set_local_address(&mut self, local_address: LocalAddress) {
        self.0.set_local_address(local_address)
    }

    pub fn tx<S: Socket, Publisher: event::EndpointPublisher>(
        &mut self,
        socket: &S,
        publisher: &mut Publisher,
    ) -> Result<usize, S::Error> {
        let mut count = 0;
        let mut entries = self.0.occupied_mut();

        for entry in entries.as_mut() {
            if let Some(remote_address) = entry.remote_address() {
                match socket.send_to(entry.payload_mut(), &remote_address) {
                    Ok(_) => {
                        count += 1;

                        publisher.on_platform_tx(event::builder::PlatformTx { count: 1 });
                    }
                    Err(err) if count > 0 && err.would_block() => {
                        break;
                    }
                    Err(err) if err.was_interrupted() || err.permission_denied() => {
                        break;
                    }
                    Err(err) => {
                        entries.finish(count);

                        publisher.on_platform_rx_error(event::builder::PlatformRxError {
                            errno: errno().0,
                        });

                        return Err(err);
                    }
                }
            }
        }

        entries.finish(count);

        Ok(count)
    }

    pub fn rx<S: Socket, Publisher: event::EndpointPublisher>(
        &mut self,
        socket: &S,
        publisher: &mut Publisher,
    ) -> Result<usize, S::Error> {
        let mut count = 0;
        let mut entries = self.0.free_mut();

        while let Some(entry) = entries.get_mut(count) {
            match socket.recv_from(entry.payload_mut()) {
                Ok((payload_len, Some(remote_address))) => {
                    entry.set_remote_address(&remote_address);
                    unsafe {
                        // Safety: The payload_len should not be bigger than the number of
                        // allocated bytes.

                        debug_assert!(payload_len < entry.payload_len());
                        let payload_len = payload_len.min(entry.payload_len());

                        entry.set_payload_len(payload_len);
                    }

                    count += 1;

                    publisher.on_platform_rx(event::builder::PlatformRx { count: 1 });
                }
                Ok((_payload_len, None)) => {}
                Err(err) if count > 0 && err.would_block() => {
                    break;
                }
                Err(err) if err.was_interrupted() => {
                    break;
                }
                Err(err) => {
                    entries.finish(count);

                    publisher
                        .on_platform_rx_error(event::builder::PlatformRxError { errno: errno().0 });

                    return Err(err);
                }
            }
        }

        entries.finish(count);

        Ok(count)
    }

    pub fn rx_queue(&mut self) -> queue::OccupiedWipe<Message> {
        self.0.occupied_wipe_mut()
    }

    pub fn tx_queue(&mut self) -> queue::Free<Message> {
        self.0.free_mut()
    }
}
