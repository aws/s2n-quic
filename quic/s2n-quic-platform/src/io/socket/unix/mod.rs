use crate::{
    buffer::Buffer as MessageBuffer,
    io::{
        rx::RxQueue as RxQueueTrait,
        socket::unix::{rx::RxBuffer, tx::TxBuffer},
        tx::{TxError, TxPayload, TxQueue as TxQueueTrait},
    },
};
use core::ops::{Deref, DerefMut};
use s2n_quic_core::{
    inet::{DatagramInfo, ExplicitCongestionNotification, SocketAddress},
    time::Timestamp,
};
use std::{io::Result as IOResult, net::ToSocketAddrs};

mod builder;
mod queue;
mod rx;
mod tx;
mod udp;

pub type RxQueue<Buffer> = RxBuffer<Buffer>;
pub type TxQueue<Buffer> = TxBuffer<Buffer>;
pub use builder::SocketBuilder;
pub use udp::UdpSocket;

#[derive(Debug)]
pub struct Socket<Tx: MessageBuffer, Rx: MessageBuffer> {
    tx_buffer: TxQueue<Tx>,
    rx_buffer: RxQueue<Rx>,
    socket: UdpSocket,
}

impl<Tx: MessageBuffer, Rx: MessageBuffer> Deref for Socket<Tx, Rx> {
    type Target = UdpSocket;

    fn deref(&self) -> &Self::Target {
        &self.socket
    }
}

impl<Tx: MessageBuffer, Rx: MessageBuffer> DerefMut for Socket<Tx, Rx> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.socket
    }
}

impl<Tx: MessageBuffer, Rx: MessageBuffer> Socket<Tx, Rx> {
    pub fn build<Addrs: ToSocketAddrs>(addr: Addrs) -> IOResult<SocketBuilder<(), ()>> {
        SocketBuilder::new(addr)
    }

    pub fn bind<Addrs: ToSocketAddrs>(addr: Addrs, tx_buffer: Tx, rx_buffer: Rx) -> IOResult<Self> {
        SocketBuilder::new(addr)?
            .tx_buffer(tx_buffer)
            .rx_buffer(rx_buffer)
            .build()
    }

    pub fn split_mut(&mut self) -> (&mut TxQueue<Tx>, &mut RxQueue<Rx>) {
        (&mut self.tx_buffer, &mut self.rx_buffer)
    }

    pub fn sync(&mut self) -> IOResult<(usize, usize)> {
        let transmissions = self.sync_tx()?;
        let receptions = self.sync_rx()?;
        Ok((transmissions, receptions))
    }

    pub fn sync_tx(&mut self) -> IOResult<usize> {
        self.tx_buffer.sync(&mut self.socket)
    }

    pub fn sync_rx(&mut self) -> IOResult<usize> {
        self.rx_buffer.sync(&mut self.socket)
    }

    #[cfg(not(feature = "mio"))]
    pub fn set_nonblocking(&mut self, nonblocking: bool) -> IOResult<()> {
        self.socket.set_nonblocking(nonblocking)
    }
}

impl<Tx: MessageBuffer, Rx: MessageBuffer> TxQueueTrait for Socket<Tx, Rx> {
    fn push<Payload: TxPayload>(
        &mut self,
        remote_address: &SocketAddress,
        ecn: ExplicitCongestionNotification,
        payload: Payload,
    ) -> Result<usize, TxError> {
        self.tx_buffer.push(remote_address, ecn, payload)
    }

    fn capacity(&self) -> usize {
        self.tx_buffer.capacity()
    }

    fn len(&self) -> usize {
        self.tx_buffer.len()
    }
}

impl<Tx: MessageBuffer, Rx: MessageBuffer> RxQueueTrait for Socket<Tx, Rx> {
    fn pop(&mut self, timestamp: Timestamp) -> Option<(DatagramInfo, &mut [u8])> {
        self.rx_buffer.pop(timestamp)
    }

    fn len(&self) -> usize {
        self.rx_buffer.len()
    }
}

#[cfg(feature = "mio")]
impl<Tx: MessageBuffer, Rx: MessageBuffer> mio::Evented for Socket<Tx, Rx> {
    fn register(
        &self,
        poll: &mio::Poll,
        token: mio::Token,
        interest: mio::Ready,
        opts: mio::PollOpt,
    ) -> IOResult<()> {
        self.socket.register(poll, token, interest, opts)
    }

    fn reregister(
        &self,
        poll: &mio::Poll,
        token: mio::Token,
        interest: mio::Ready,
        opts: mio::PollOpt,
    ) -> IOResult<()> {
        self.socket.reregister(poll, token, interest, opts)
    }

    fn deregister(&self, poll: &mio::Poll) -> IOResult<()> {
        self.socket.deregister(poll)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::buffer::VecBuffer;

    fn test_socket<Addrs: ToSocketAddrs + Copy>(addr: Addrs) {
        let bind = |slot_count| -> IOResult<Socket<_, _>> {
            let slot_size = 1200;
            let tx_buffer = VecBuffer::new(slot_count, slot_size);
            let rx_buffer = VecBuffer::new(slot_count, slot_size);

            Socket::bind(addr, tx_buffer, rx_buffer)
        };

        let slot_count = 16;

        let mut server = bind(slot_count).unwrap();
        let server_addr: SocketAddress = server.local_addr().unwrap().into();
        let mut client = bind(slot_count).unwrap();
        let client_addr = client.local_addr().unwrap();

        while client.can_push() {
            client
                .push_encoder_value(&server_addr, Default::default(), 123u64)
                .unwrap();
        }

        while !TxQueueTrait::is_empty(&client) {
            client.sync_tx().unwrap();
        }

        let mut count = 0;
        let mut failure_count = 0;
        while count < slot_count {
            if let Err(err) = server.sync_rx() {
                if err.kind() == std::io::ErrorKind::WouldBlock {
                    std::thread::sleep(core::time::Duration::from_millis(1));
                    assert!(failure_count < 10, "blocked too many times");
                    failure_count += 1;
                    continue;
                }
                panic!("socket rx failed: {:?}", err);
            }

            let (tx, rx) = server.split_mut();
            while let Some((info, payload)) = rx.pop(crate::time::now()) {
                assert_eq!(info.remote_address.port(), client_addr.port());

                let payload = s2n_codec::DecoderBufferMut::new(payload);
                let (value, remaining) = payload.decode::<u64>().unwrap();
                assert_eq!(value, 123);
                assert!(remaining.is_empty());

                tx.push_encoder_value(&info.remote_address, Default::default(), value)
                    .unwrap();

                count += 1;
            }
        }

        assert_eq!(count, slot_count);
    }

    #[test]
    fn ipv4_test() {
        test_socket("0.0.0.0:0");
    }

    #[cfg(feature = "ipv6")]
    #[test]
    fn ipv6_test() {
        test_socket("[::1]:0");
    }
}
