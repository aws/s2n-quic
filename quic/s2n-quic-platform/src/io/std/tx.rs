use crate::{
    buffer,
    message::{
        queue,
        simple::{Message, Ring},
        Message as MessageTrait,
    },
    socket,
};
use s2n_quic_core::io::tx;
use std::io;

impl_io!(Tx);
impl_socket_raw_delegate!(
    impl[Buffer: buffer::Buffer, Socket: socket::raw::AsRaw] Tx<Buffer, Socket>,
    |self| &self.socket
);
impl_socket_mio_delegate!(
    impl[Buffer: buffer::Buffer, Socket: mio::Evented] Tx<Buffer, Socket>,
    |self| &self.socket
);

impl<
        'a,
        Buffer: buffer::Buffer,
        Socket: socket::Simple<Error = io::Error> + socket::Socket<Error = io::Error>,
    > tx::Tx<'a> for Tx<Buffer, Socket>
{
    type Queue = queue::Free<'a, Message>;
    type Error = io::Error;

    fn queue(&'a mut self) -> Self::Queue {
        self.queue.free_mut()
    }

    fn len(&self) -> usize {
        self.queue.occupied_len()
    }

    fn poll_transmit(
        &mut self,
        cx: &mut core::task::Context<'_>,
    ) -> core::task::Poll<io::Result<usize>> {
        let Self { socket, queue } = self;
        socket.poll_transmit(cx, |socket| {
            let mut count = 0;
            let mut occupied = queue.occupied_mut();

            for entry in occupied.as_mut() {
                if let Some(remote_address) = entry.remote_address() {
                    match socket.send_to(entry.payload_mut(), &remote_address) {
                        Ok(_) => {
                            count += 1;
                        }
                        Err(err) => {
                            if count > 0 && err.kind() == io::ErrorKind::WouldBlock {
                                break;
                            } else {
                                occupied.finish(count);
                                return Err(err);
                            }
                        }
                    }
                }
            }

            occupied.finish(count);

            Ok(count)
        })
    }
}

#[cfg(test)]
mod tests {
    impl_io_tx_tests!();
}
