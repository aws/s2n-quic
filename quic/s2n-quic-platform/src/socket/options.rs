// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::syscall;
use s2n_quic_core::inet::SocketAddress;
use std::{
    io,
    net::{SocketAddr, TcpListener, UdpSocket},
};

#[derive(Clone, Copy, Debug, Default)]
pub enum ReusePort {
    #[default]
    Disabled,
    /// Enables reuse port before binding the socket
    ///
    /// NOTE: the provided `addr` must not be bound to a random port (`0`)
    BeforeBind,
    /// Enables reuse port after binding the socket
    AfterBind,
}

#[derive(Clone, Debug)]
#[non_exhaustive]
pub struct Options {
    pub addr: SocketAddr,
    pub reuse_address: bool,
    pub reuse_port: ReusePort,
    pub gro: bool,
    pub blocking: bool,
    pub delay: bool,
    pub send_buffer: Option<usize>,
    pub recv_buffer: Option<usize>,
    pub backlog: usize,
}

impl Default for Options {
    #[inline]
    fn default() -> Self {
        Self {
            addr: SocketAddress::default().into(),
            reuse_address: false,
            reuse_port: Default::default(),
            gro: true,
            blocking: false,
            send_buffer: None,
            recv_buffer: None,
            delay: false,
            backlog: 4096,
        }
    }
}

impl Options {
    #[inline]
    pub fn new(addr: SocketAddr) -> Self {
        Self {
            addr,
            ..Default::default()
        }
    }

    #[inline]
    pub fn build_udp(&self) -> io::Result<UdpSocket> {
        let socket = syscall::udp_socket(self.addr)?;

        if self.gro {
            let _ = syscall::configure_gro(&socket);
        }

        let _ = syscall::configure_tos(&socket);
        let _ = syscall::configure_mtu_disc(&socket);

        self.build_common(&socket)?;

        let socket = socket.into();
        Ok(socket)
    }

    #[inline]
    pub fn build_tcp_listener(&self) -> io::Result<TcpListener> {
        let domain = socket2::Domain::for_address(self.addr);
        let ty = socket2::Type::STREAM;
        let protocol = socket2::Protocol::TCP;

        let socket = socket2::Socket::new(domain, ty, Some(protocol))?;

        socket.set_nodelay(!self.delay)?;

        self.build_common(&socket)?;

        socket.listen(self.backlog.try_into().unwrap_or(core::ffi::c_int::MAX))?;

        Ok(socket.into())
    }

    fn build_common(&self, socket: &socket2::Socket) -> io::Result<()> {
        socket.set_reuse_address(self.reuse_address)?;
        socket.set_nonblocking(!self.blocking)?;

        if let Some(send_buffer) = self.send_buffer {
            let _ = socket.set_send_buffer_size(send_buffer);
        }

        if let Some(recv_buffer) = self.recv_buffer {
            let _ = socket.set_recv_buffer_size(recv_buffer);
        }

        if let ReusePort::BeforeBind = self.reuse_port {
            assert_ne!(self.addr.port(), 0);
            set_reuse_port(socket)?;
        }

        socket.bind(&self.addr.into())?;

        if let ReusePort::AfterBind = self.reuse_port {
            set_reuse_port(socket)?;
        }

        Ok(())
    }
}

#[cfg(windows)]
fn set_reuse_port(_socket: &socket2::Socket) -> io::Result<()> {
    Err(io::Error::new(
        io::ErrorKind::InvalidInput,
        "reuse port is not supported on windows",
    ))
}

#[cfg(not(windows))]
fn set_reuse_port(socket: &socket2::Socket) -> io::Result<()> {
    socket.set_reuse_port(true)
}
