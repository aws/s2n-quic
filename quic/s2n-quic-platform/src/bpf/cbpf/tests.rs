// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use crate::socket::options::{Options, ReusePort};
use std::{
    io::{self, Write},
    net::UdpSocket,
};

macro_rules! test {
    ($name:ident,no_run, $asm:expr, $instructions:expr $(,)?) => {
        #[test]
        fn $name() {
            Test {
                name: stringify!($name),
                asm: $asm,
                instructions: $instructions,
                checks: Checks {
                    run: false,
                    ..Default::default()
                },
            }
            .check();
        }
    };
    ($name:ident, $asm:expr, $instructions:expr $(,)?) => {
        #[test]
        fn $name() {
            Test {
                name: stringify!($name),
                asm: $asm,
                instructions: $instructions,
                checks: Default::default(),
            }
            .check();
        }
    };
}

test!(
    example_ipv4_tcp_packets,
    r#"
    ldh [12]
    jne #0x800, drop
    ldb [23]
    jneq #6, drop
    ret #-1
    drop: ret #0
    "#,
    &[
        ldh(abs(12)),
        jneq(0x800, 3, 0),
        ldb(abs(23)),
        jneq(6, 1, 0),
        ret(u32::MAX),
        ret(0)
    ],
);

test!(
    example_interface_index_13,
    r#"
    ld ifidx
    jneq #13, drop
    ret #-1
    drop: ret #0
    "#,
    &[
        ld(ancillary::skb::ifindex()),
        jneq(13, 1, 0),
        ret(u32::MAX),
        ret(0),
    ],
);

test!(
    example_vlan_w_id_10,
    r#"
    ld vlan_tci
    jneq #10, drop
    ret #-1
    drop: ret #0
    "#,
    &[
        ld(ancillary::skb::vlan_tci()),
        jneq(10, 1, 0),
        ret(u32::MAX),
        ret(0),
    ],
);

const FIRST_BYTE_MAX: u8 = 0b0011_1111;

test!(
    first_byte_routing,
    &format!(
        r#"
    ldb [0]
    and #{}
    jneq #{}, reader
    ret #0
    reader: ret #1
    "#,
        !FIRST_BYTE_MAX,
        FIRST_BYTE_MAX + 1
    ),
    &[
        // load the first byte of the packet
        ldb(abs(0)),
        // mask off the LSBs
        and(!FIRST_BYTE_MAX as _),
        // IF:
        // the control bit is set
        jneq(FIRST_BYTE_MAX as u32 + 1, 1, 0),
        // THEN:
        // return a 0 indicating we want to route to the writer socket
        ret(0),
        // ELSE:
        // return a 1 indicating we want to route to the reader socket
        ret(1),
    ],
);

test!(
    hash_routing,
    r#"
    ld rxhash
    and #0x1
    ret %a
    "#,
    &[ld(ancillary::skb::hash()), and(1), ret_a(),]
);

test!(
    len_check,
    r#"
    ld len
    jgt #100, drop
    ret #0
    drop: ret #1
    "#,
    &[ld(len()), jgt(100, 1, 0), ret(0), ret(1)]
);

test!(
    comparisons,
    r#"
    ld [0]
    jgt #1, drop
    jle #2, drop
    jeq #3, drop
    jneq #4, drop
    drop: ret #0
    "#,
    &[
        ld(abs(0)),
        jgt(1, 3, 0),
        jle(2, 2, 0),
        jeq(3, 1, 0),
        jneq(4, 0, 0),
        ret(0)
    ]
);

// pkt += match *pkt >> 6 {
//     0 => 1,
//     1 => 2,
//     2 => 4,
//     3 => 8,
// };
test!(
    varint_skip,
    r#"
        ldx #0                        ;; initialize the cursor

    varint_skip:
        ldb [x + 0]                   ;; load the current byte from the cursor
        rsh #6                        ;; shift the first byte into the 2 bit mask
        add #1                        ;; add 1 to the mask
        jgt #2, varint_skip_done      ;; if the mask is 1 or 2, we're done

        add #1                        ;; increment by 1 - either it's 4 or 8
        jeq #4, varint_skip_done      ;; if the mask is 4, we're done

        add #4                        ;; increment by 4, getting us to 8

    varint_skip_done:
        add %x                        ;; add the varint length to the cursor
        tax                           ;; put the new cursor into x
        ret #0                        ;; return 0 just for the test
    "#,
    &[
        // initialize the cursor
        ldx(imm(0)),
        // load the current byte from the cursor
        ldb(ind(0)),
        // shift the first byte into the 2 bit mask
        rsh(6),
        // add 1 to the mask
        add(1),
        // if the mask is 1 or 2, we're done
        jgt(2, 3, 0),
        // increment by 1 - either it's 4 or 8
        add(1),
        // if the mask is 4, we're done
        jeq(4, 1, 0),
        // increment by 4, getting us to 8
        add(4),
        // add the varint length to the cursor
        add_x(),
        // put the new cursor into x
        tax(),
        // return 0 just for the test
        ret(0),
    ]
);

struct Test<'a> {
    name: &'a str,
    asm: &'a str,
    instructions: &'a [Instruction],
    checks: Checks,
}

impl Test<'_> {
    fn check(&self) {
        let prog = Program::new(self.instructions);

        insta::assert_snapshot!(format!("{}_display", self.name), prog);
        insta::assert_debug_snapshot!(format!("{}_debug", self.name), prog);
        let mut actual_c_literal = String::new();
        for i in self.instructions {
            use core::fmt::Write;
            writeln!(actual_c_literal, "{i:#},").unwrap();
        }
        insta::assert_snapshot!(format!("{}_c_literal", self.name), actual_c_literal);

        self.checks.compile(
            &[
                // this ensures the instruction functions actually map to the correct asm
                self.asm,
                // this ensures that the disassembly can reproduce the original asm code
                &prog.to_string(),
            ],
            &actual_c_literal,
        );
        self.checks.run(&prog);
    }
}

struct BpfAsm {
    path: String,
}

impl BpfAsm {
    /// Tries to resolve a `bpf_asm` binary capable of assembling BPF programs.
    ///
    /// If none is found, then `None` is returned.
    fn resolve() -> Option<&'static Self> {
        use std::sync::OnceLock;

        static INSTANCE: OnceLock<Option<BpfAsm>> = OnceLock::new();

        INSTANCE
            .get_or_init(|| {
                let bpf_asm = std::process::Command::new("which")
                    .arg("bpf_asm")
                    .output()
                    .ok()?;

                let bpf_asm = core::str::from_utf8(&bpf_asm.stdout).unwrap_or("").trim();
                if bpf_asm.is_empty() {
                    return None;
                }

                Some(Self {
                    path: bpf_asm.to_string(),
                })
            })
            .as_ref()
    }

    fn assemble(&self, asm: &str) -> String {
        let mut file = tempfile::NamedTempFile::new().unwrap();
        file.write_all(asm.as_bytes()).unwrap();

        let output = std::process::Command::new(&self.path)
            .arg("-c")
            .arg(file.path())
            .output()
            .unwrap();

        assert!(output.status.success(), "{output:?}\nINPUT:\n{asm}");
        let output = core::str::from_utf8(&output.stdout).unwrap();

        output.to_string()
    }
}

#[derive(Debug)]
struct Checks {
    compile: bool,
    run: bool,
}

impl Default for Checks {
    fn default() -> Self {
        Self {
            compile: true,
            run: true,
        }
    }
}

impl Checks {
    fn compile(&self, asm: &[&str], expected: &str) {
        if !self.compile {
            return;
        }
        let Some(bpf_asm) = BpfAsm::resolve() else {
            eprintln!("`bpf_asm` not found in environment - skipping tests");
            return;
        };

        for asm in asm {
            let out = bpf_asm.assemble(asm);
            assert_eq!(
                expected, out,
                "\nINPUT:\n{asm}\nBPF_OUT:\n{out}\nEXPECTED:\n{expected}\n"
            );
        }
    }

    fn run(&self, prog: &Program) {
        if !self.run {
            return;
        }

        use std::sync::OnceLock;

        static PAIR: OnceLock<io::Result<(UdpSocket, UdpSocket)>> = OnceLock::new();

        let pair = PAIR.get_or_init(|| {
            udp_pair().and_then(|(a, b)| {
                probe(&a)?;
                probe(&b)?;
                Ok((a, b))
            })
        });

        let Ok((a, b)) = pair else {
            eprintln!("skipping tests due to environment not supporting bpf programs");
            return;
        };

        for socket in [a, b] {
            prog.attach(socket).unwrap();
        }
    }
}

fn udp_pair() -> io::Result<(UdpSocket, UdpSocket)> {
    let mut options = Options {
        gro: false,
        blocking: false,
        // set the reuse port option after binding to avoid port collisions
        reuse_port: ReusePort::AfterBind,
        ..Default::default()
    };

    let writer = options.build_udp()?;

    // bind the sockets to the same address
    options.addr = writer.local_addr()?;
    // now that we have a concrete port from the OS, we set the option before the bind call
    options.reuse_port = ReusePort::BeforeBind;

    let reader = options.build_udp()?;

    Ok((writer, reader))
}

/// Probe for BPF support in the environment
fn probe<F: std::os::fd::AsRawFd>(socket: &F) -> io::Result<()> {
    use libc::{sock_filter, sock_fprog};

    // $ echo 'ret #0' | bfp_asm -c
    // { 0x06,  0,  0, 0000000000 },
    let instructions = [sock_filter {
        code: 0x06,
        jt: 0,
        jf: 0,
        k: 0,
    }];

    let prog = sock_fprog {
        filter: instructions.as_ptr() as *const _ as *mut _,
        len: instructions.len() as _,
    };

    let ret = unsafe {
        libc::setsockopt(
            socket.as_raw_fd(),
            libc::SOL_SOCKET,
            <Cbpf as super::super::instruction::Dialect>::SOCKOPT as _,
            &prog as *const _ as *const _,
            core::mem::size_of::<sock_fprog>() as _,
        )
    };

    if ret < 0 {
        Err(io::Error::last_os_error())
    } else {
        Ok(())
    }
}
