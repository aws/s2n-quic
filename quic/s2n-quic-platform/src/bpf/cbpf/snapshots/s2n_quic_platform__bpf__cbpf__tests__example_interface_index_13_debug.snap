---
source: quic/s2n-quic-platform/src/bpf/cbpf/tests.rs
expression: prog
---
Program {
    instructions: [
        Instruction {
            code: 32,
            jt: 0,
            jf: 0,
            class: LD,
            size: W,
            mode: ABS,
            k: 0xfffff008,
        },
        Instruction {
            code: 21,
            jt: 0,
            jf: 1,
            class: JMP,
            op: JEQ,
            jf: 1,
            k: 0xd,
        },
        Instruction {
            code: 6,
            jt: 0,
            jf: 0,
            class: RET,
            size: W,
            mode: IMM,
            k: 0xffffffff,
        },
        Instruction {
            code: 6,
            jt: 0,
            jf: 0,
            class: RET,
            size: W,
            mode: IMM,
            k: 0,
        },
    ],
}
